File: src/lib/math/vector.ts
import { EulerRotation, QuaternionRotation, Vector3 } from './common';

export class V3 {
  static all(value: number): Vector3 {
    return [value, value, value];
  }

  static sum(a: Vector3, b: Vector3): Vector3 {
    return [a[0] + b[0], a[1] + b[1], a[2] + b[2]];
  }

  static sub(a: Vector3, b: Vector3): Vector3 {
    return [a[0] - b[0], a[1] - b[1], a[2] - b[2]];
  }

  static mean(a: Vector3, b: Vector3): Vector3 {
    return [
      (a[0] + b[0]) / 2,
      (a[1] + b[1]) / 2,
      (a[2] + b[2]) / 2,
    ];
  }

  static negative(a: Vector3): Vector3 {
    return [-a[0], -a[1], -a[2]];
  }

  static scale(a: Vector3, b: number): Vector3 {
    return [a[0] * b, a[1] * b, a[2] * b];
  }

  static mul(a: Vector3, b: Vector3): Vector3 {
    return [a[0] * b[0], a[1] * b[1], a[2] * b[2]];
  }

  static rotate(v: Vector3, eulerAngleRadians: Vector3): Vector3 {
    return rotateVector(v, eulerAngleRadians);
  }

  static normalize(v: Vector3): Vector3 {
    const mag = Math.hypot(v[0], v[1], v[2]);
    return [v[0] / mag, v[1] / mag, v[2] / mag];
  }

  static min(a: Vector3, b: Vector3): Vector3 {
    return [Math.min(a[0], b[0]), Math.min(a[1], b[1]), Math.min(a[2], b[2])];
  }

  static max(a: Vector3, b: Vector3): Vector3 {
    return [Math.max(a[0], b[0]), Math.max(a[1], b[1]), Math.max(a[2], b[2])];
  }

  static distance(a: Vector3, b: Vector3): number {
    return Math.hypot(a[0] - b[0], a[1] - b[1], a[2] - b[2]);
  }

  static lerp(v0: Vector3, v1: Vector3, t: number): Vector3 {
    return [
      v0[0] + t * (v1[0] - v0[0]),
      v0[1] + t * (v1[1] - v0[1]),
      v0[2] + t * (v1[2] - v0[2]),
    ];
  }

  static lerpScalar(v0: number, v1: number, t: number): number {
    return v0 + t * (v1 - v0);
  }

  // eslint-disable-next-line
  static noInterp<T extends number[]>(v0: T, _v1: T, _t: number): T {
    return [...v0] as T;
  }

  static slerp(q0: QuaternionRotation, q1: QuaternionRotation, t: number): QuaternionRotation {
    // Compute the dot product (cosine of the angle between the quaternions)
    let dot = q0[0] * q1[0] + q0[1] * q1[1] + q0[2] * q1[2] + q0[3] * q1[3];
    dot = Math.max(-1.0, Math.min(1.0, dot));

    // If the dot product is negative, the quaternions have opposite handed-ness
    // and slerp won't take the shorter path. So we negate one quaternion.
    if (dot < 0.0) {
      // eslint-disable-next-line no-param-reassign
      q1 = [-q1[0], -q1[1], -q1[2], -q1[3]];
      dot = -dot;
    }

    const DOT_THRESHOLD = 0.9995;
    if (dot > DOT_THRESHOLD) {
      // If the quaternions are nearly parallel, use linear interpolation
      // to avoid division by 0
      const result = [
        q0[0] + t * (q1[0] - q0[0]),
        q0[1] + t * (q1[1] - q0[1]),
        q0[2] + t * (q1[2] - q0[2]),
        q0[3] + t * (q1[3] - q0[3]),
      ];
      // Normalize the result
      const mag = Math.sqrt(result[0] * result[0] + result[1] * result[1] + result[2] * result[2] + result[3] * result[3]);
      return [result[0] / mag, result[1] / mag, result[2] / mag, result[3] / mag];
    }

    // Calculate the angle between the quaternions
    const theta_0 = Math.acos(dot); // θ₀ = angle between input quaternions
    const theta = theta_0 * t; // theta = angle between q0 and the result
    const sin_theta = Math.sin(theta); // Compute this once
    const sin_theta_0 = Math.sin(theta_0); // Compute this once

    const s0 = Math.cos(theta) - dot * sin_theta / sin_theta_0; // == sin(theta_0 - theta) / sin(theta_0)
    const s1 = sin_theta / sin_theta_0;

    return [
      (s0 * q0[0]) + (s1 * q1[0]),
      (s0 * q0[1]) + (s1 * q1[1]),
      (s0 * q0[2]) + (s1 * q1[2]),
      (s0 * q0[3]) + (s1 * q1[3]),
    ];
  }
}

export function rotateVector(
  position: Vector3,
  angle: EulerRotation,
): Vector3 {
  const [x, y, z] = position;
  const rx = angle[0];
  const ry = angle[1];
  const rz = angle[2];

  // Rotation matrices
  const R_x = [
    [1, 0, 0],
    [0, Math.cos(rx), -Math.sin(rx)],
    [0, Math.sin(rx), Math.cos(rx)],
  ];

  const R_y = [
    [Math.cos(ry), 0, Math.sin(ry)],
    [0, 1, 0],
    [-Math.sin(ry), 0, Math.cos(ry)],
  ];

  const R_z = [
    [Math.cos(rz), -Math.sin(rz), 0],
    [Math.sin(rz), Math.cos(rz), 0],
    [0, 0, 1],
  ];

  // Helper function to multiply matrices
  function multiplyMatrixVector(matrix: number[][], vector: Vector3): Vector3 {
    return [
      matrix[0][0] * vector[0] + matrix[0][1] * vector[1] + matrix[0][2] * vector[2],
      matrix[1][0] * vector[0] + matrix[1][1] * vector[1] + matrix[1][2] * vector[2],
      matrix[2][0] * vector[0] + matrix[2][1] * vector[1] + matrix[2][2] * vector[2],
    ];
  }

  // Combine rotations
  const intermediateVector1 = multiplyMatrixVector(R_x, [x, y, z]);
  const intermediateVector2 = multiplyMatrixVector(R_y, intermediateVector1);
  const rotatedVector = multiplyMatrixVector(R_z, intermediateVector2);

  return rotatedVector;
}

File: golang/internal/math/vector.go
package math

import "math"

func V3All(value float64) Vector3 {
	return Vector3{value, value, value}
}

func V3Sum(a, b Vector3) Vector3 {
	return Vector3{a[0] + b[0], a[1] + b[1], a[2] + b[2]}
}

func V3Sub(a, b Vector3) Vector3 {
	return Vector3{a[0] - b[0], a[1] - b[1], a[2] - b[2]}
}

func V3Mean(a, b Vector3) Vector3 {
	return Vector3{(a[0] + b[0]) / 2, (a[1] + b[1]) / 2, (a[2] + b[2]) / 2}
}

func V3Negative(a Vector3) Vector3 {
	return Vector3{-a[0], -a[1], -a[2]}
}

func V3Scale(a Vector3, b float64) Vector3 {
	return Vector3{a[0] * b, a[1] * b, a[2] * b}
}

func V3Mul(a, b Vector3) Vector3 {
	return Vector3{a[0] * b[0], a[1] * b[1], a[2] * b[2]}
}

func V3Rotate(v Vector3, eulerAngleRadians Vector3) Vector3 {
	return RotateVector(v, eulerAngleRadians)
}

func V3Normalize(v Vector3) Vector3 {
	mag := math.Sqrt(v[0]*v[0] + v[1]*v[1] + v[2]*v[2])
	if mag == 0 {
		return Vector3{}
	}
	return Vector3{v[0] / mag, v[1] / mag, v[2] / mag}
}

func V3Dot(a, b Vector3) float64 {
	return a[0]*b[0] + a[1]*b[1] + a[2]*b[2]
}

func V3Min(a, b Vector3) Vector3 {
	return Vector3{math.Min(a[0], b[0]), math.Min(a[1], b[1]), math.Min(a[2], b[2])}
}

func V3Max(a, b Vector3) Vector3 {
	return Vector3{math.Max(a[0], b[0]), math.Max(a[1], b[1]), math.Max(a[2], b[2])}
}

func V3Distance(a, b Vector3) float64 {
	dx, dy, dz := a[0]-b[0], a[1]-b[1], a[2]-b[2]
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

func V3Lerp(v0, v1 Vector3, t float64) Vector3 {
	return Vector3{
		v0[0] + t*(v1[0]-v0[0]),
		v0[1] + t*(v1[1]-v0[1]),
		v0[2] + t*(v1[2]-v0[2]),
	}
}

func V3LerpScalar(v0, v1, t float64) float64 {
	return v0 + t*(v1-v0)
}

func V3NoInterpScalar(v0, _, _ float64) float64 {
	return v0
}

func V3NoInterp[T ~[]float64](v0 T, _, _ float64) T {
	out := make(T, len(v0))
	copy(out, v0)
	return out
}

func V3Slerp(q0, q1 QuaternionRotation, t float64) QuaternionRotation {
	dot := q0[0]*q1[0] + q0[1]*q1[1] + q0[2]*q1[2] + q0[3]*q1[3]
	dot = math.Max(-1.0, math.Min(1.0, dot))

	if dot < 0.0 {
		q1 = QuaternionRotation{-q1[0], -q1[1], -q1[2], -q1[3]}
		dot = -dot
	}

	const dotThreshold = 0.9995
	if dot > dotThreshold {
		result := QuaternionRotation{
			q0[0] + t*(q1[0]-q0[0]),
			q0[1] + t*(q1[1]-q0[1]),
			q0[2] + t*(q1[2]-q0[2]),
			q0[3] + t*(q1[3]-q0[3]),
		}
		mag := math.Sqrt(result[0]*result[0] + result[1]*result[1] + result[2]*result[2] + result[3]*result[3])
		return QuaternionRotation{result[0] / mag, result[1] / mag, result[2] / mag, result[3] / mag}
	}

	theta0 := math.Acos(dot)
	theta := theta0 * t
	sinTheta := math.Sin(theta)
	sinTheta0 := math.Sin(theta0)

	s0 := math.Cos(theta) - dot*sinTheta/sinTheta0
	s1 := sinTheta / sinTheta0

	return QuaternionRotation{
		(s0 * q0[0]) + (s1 * q1[0]),
		(s0 * q0[1]) + (s1 * q1[1]),
		(s0 * q0[2]) + (s1 * q1[2]),
		(s0 * q0[3]) + (s1 * q1[3]),
	}
}

func RotateVector(position Vector3, angle EulerRotation) Vector3 {
	x, y, z := position[0], position[1], position[2]
	rx, ry, rz := angle[0], angle[1], angle[2]

	rxM := [3][3]float64{{1, 0, 0}, {0, math.Cos(rx), -math.Sin(rx)}, {0, math.Sin(rx), math.Cos(rx)}}
	ryM := [3][3]float64{{math.Cos(ry), 0, math.Sin(ry)}, {0, 1, 0}, {-math.Sin(ry), 0, math.Cos(ry)}}
	rzM := [3][3]float64{{math.Cos(rz), -math.Sin(rz), 0}, {math.Sin(rz), math.Cos(rz), 0}, {0, 0, 1}}

	multiply := func(matrix [3][3]float64, vector Vector3) Vector3 {
		return Vector3{
			matrix[0][0]*vector[0] + matrix[0][1]*vector[1] + matrix[0][2]*vector[2],
			matrix[1][0]*vector[0] + matrix[1][1]*vector[1] + matrix[1][2]*vector[2],
			matrix[2][0]*vector[0] + matrix[2][1]*vector[1] + matrix[2][2]*vector[2],
		}
	}

	v1 := multiply(rxM, Vector3{x, y, z})
	v2 := multiply(ryM, v1)
	return multiply(rzM, v2)
}
